routing/payment_lifecycle.go

```diff
@@ -24,14 +26,85 @@ type paymentLifecycle struct {
 	currentHeight int32
 }
 
+// paymentState uses the passed payment to find the latest information we need
+// to act on every iteration of the payment loop.
+func (p *paymentLifecycle) paymentState(payment *channeldb.MPPayment) (
+	uint32, lnwire.MilliSatoshi, lnwire.MilliSatoshi, bool, error) {
+
+	var (
+		// Keep track of the number of active shards we have
+		// in-flight...
+		activeShards uint32
+
+		// ...and the total value and fees of all in-flight and
+		// settled shards.
+		remValue = p.totalAmount
+		feeLimit = p.feeLimit
+
+		// If the payment is already marked failed, we'll get
+		// ready to terminate the moment all shards have
+		// returned.
+		terminate = payment.FailureReason != nil
+	)
+
+	// Go through the HTLCs for this payment, determining if there
+	// are any in flight.
+	for _, a := range payment.HTLCs {
+		// This HTLC already failed, ignore.
+		if a.Failure != nil {
+			continue
+		}
+
+		// Subtract settled or in flight attempts from the
+		// remaining value.
+		amt := a.Route.ReceiverAmt()
+		if amt > remValue {
+			return 0, 0, 0, false, fmt.Errorf("amount "+
+				"sent %v exceeds remaining amount %v",
+				amt, remValue)
+		}
+		remValue -= amt
+
+		fee := a.Route.TotalFees()
+		if fee > feeLimit {
```

Comment:
Add comment that this should never happen?

Reply:
Not applicable anymore. It can happen 😱